### PR TITLE
[Hackathon] Project NEXUS-SBERT: Dendritic Optimization for Privacy-First Edge RAG

### PR DESCRIPTION
**Pull request by aakanksha-singh-hub**
Authors: Aakanksha Singh, Mihir Phalke

**Summary**
This PR introduces Project NEXUS-SBERT, which applies Perforated AI’s dendritic architecture to the widely used sentence-transformers/all-MiniLM-L6-v2 model (50M+ monthly downloads).
The work demonstrates stable fine-tuning and improved convergence for privacy-first Retrieval-Augmented Generation (RAG) use cases, particularly on edge and resource-constrained deployments.

**Key Results**
- Improved Convergence: Final loss of 0.0036 vs 0.0038 baseline (~5.3% error reduction)
- Training Stability: Reduced overfitting and avoided catastrophic forgetting through controlled dendritic switching
- Minimal Overhead: ~0.6% model size increase (≈587 KB)
- High Practical Relevance: Target model is an industry-standard embedding backbone for RAG systems

**Code Contributions**
- Added a configurable --warmup_epochs parameter to the training pipeline
- This prevents premature dendritic activation and significantly stabilizes early training
- The feature is reusable and available for the broader Perforated AI community

**Experiments & Analysis**
- W&B Bayesian Sweep: [Live dashboard with parallel-coordinates visualization](https://wandb.ai/aakanksha-singh0205-kj-somaiya-school-of-engineering/PerforatedAI-Examples_hackathonProjects_Project-Nexus-SBERT_src/sweeps/sv638g8j)
- Business Case: Cost analysis estimating ~$52K/year savings for privacy-compliant RAG deployments
- Documentation: Detailed README, case study, and sweep report included

**Visual Evidence**
- PAI.png: Included in assets/PAI.png, illustrating dendritic activation behavior during training
- Sweep Dashboard: Linked above for full hyperparameter and convergence analysis

**Documentation**
- README.md – Setup instructions and results overview
- docs/CASE_STUDY.md – Technical narrative and business impact
- docs/SWEEP_REPORT.md – Bayesian sweep methodology and findings

**Project Location**
`Examples/hackathonProjects/Project-Nexus-SBERT/`

**Reproducibility**
```
# Install dependencies
pip install -r requirements.txt

# Run dendritic training
python src/train_nexus.py --use_dendrites --epochs 10 --save_dir experiments/dendritic

# Run baseline for comparison
python src/train_nexus.py --epochs 10 --save_dir experiments/baseline
```

### CODE REVIEW

**Comment by RorryB** [comment]
This project looks great!  Two notes:

1 - I think the wandb sweep may have been the wrong link.  When I click it I am shown this sweep with only 2 runs, both of which crashed:
image

2 - in initialize_pai, you use save_name="PAI".  This is the default for just regular runs, but when running a sweep it will give all of the runs the same save_name, causing them to overwrite each other.  This means the PAI/PAI.png that you included is just the final one rather than the best one.  If possible, would you be able to re-run the sweep, while iterating on the save_name so each run is saved differently?  or alternatively, would you be able to just do a single run with the optimal parameters?

**Comment by aakanksha-singh-hub** [comment]
Thank you so so much for going through our project, and giving us your time and your valuable feedback. I've addressed both points:



**Fixed WandB Link:** Apologies for the broken link. I've initialized a fresh sweep and updated SWEEP_REPORT.md with the correct [URL](https://wandb.ai/aakanksha-singh0205-kj-somaiya-school-of-engineering/PerforatedAI-Examples_hackathonProjects_Project-Nexus-SBERT_src/sweeps/lzwt00l9). The sweep has just begun, and will be populated with the correct results in a few hours from my end.

**Artifact Overwriting:** 
You were absolutely right about `initialize_pai` overwriting files, thank you so much for pointing this out. I've updated the code to use a dynamic `save_name` (appending the unique run ID), so each run in the sweep will now save its own distinct artifacts (e.g., PAI_lzwt00l9.png) without conflict.

**Comment by RorryB** [comment]
Sounds great!  Once the sweep concludes please update the PAI.png with the one from the optimal run and I'll take a look at the new graph and the wandb sweep.

**Comment by aakanksha-singh-hub** [comment]
Hi!
The sweep has concluded successfully! I've updated [PAI/PAI.png](https://github.com/aakanksha-singh-hub/PerforatedAI/blob/submission/project-nexus/Examples/hackathonProjects/Project-Nexus-SBERT/PAI/PAI.png) with the graph from the optimal run (graceful-sweep-11: lr=2e-5, batch_size=32, warmup_epochs=4).

Results:
- Best Spearman: 0.89167 (2.6% error reduction over baseline)
- Training Efficiency: 8 epochs vs 10 baseline (40% faster)
- Dendritic Activation: 2 successful architectural switches visible in the graph
- W&B Links:
1. Comprehensive Report: https://wandb.ai/aakanksha-singh0205-kj-somaiya-school-of-engineering/PerforatedAI-Examples_hackathonProjects_Project-Nexus-SBERT_src/reports/Project-NEXUS-Hyperparameter-Sweep-for-Dendritic-Sentence-BERT--VmlldzoxNTU5MzkzMw
2. Sweep Dashboard: https://wandb.ai/aakanksha-singh0205-kj-somaiya-school-of-engineering/PerforatedAI-Examples_hackathonProjects_Project-Nexus-SBERT_src/sweeps/i31nvl4m
All 12 runs successfully activated dendrites and achieved >0.88 Spearman, demonstrating robust performance across hyperparameter choices.

Ready for your review! Thank you so much Rorry for giving us your time

**Comment by RorryB** [comment]
Hey, I spent some time tonight running your code.  Everything looks correct, i.e. nothing wrong with the way you implemented things.  But I think you're running into [graph problem 4](https://docs.google.com/document/d/1HygopGvDopYEF_rBlQvSbifgK-3GgQqnerX7yVngvHs/edit?tab=t.0) of overfitting here.  I made a small update to your code so add_extra_score adds the trianing score rather than training loss.  that way you can see it much easier on the generated graph:
PAI_i130807b_14 01 2026 21 51 59_noImprove_lr_0

ran with the reccomended settings, which for me actually ended up deleting the dendrites at the end, showing they did not actually end up helping:

python src/train_nexus.py --use_dendrites --epochs 8 --batch_size 32 --lr 2e-5 --warmup_epochs 4 --save_dir experiments/dendritic

With alternative settings you can really see the overfitting more strongly:



I also tried a handful of other settings to try to counteract the overfitting but didn't have any luck on my side.  Checking in my additional choices file as well in case you want to experiment with that as well.  The next things I would try if I was you is 1 - try [wrapping different modules](https://github.com/PerforatedAI/PerforatedAI/blob/main/API/customization.md#21---setting-which-modules-to-use-for-dendrite-learning) instead of just the dense block.  I can help with this if you need support with the documentation.  
2 - try doing a compression experiment instead.  I am seeing pretty consistently that dendrites are spiking training scores.  This can sometimes be a sign that the systems are compatible, but the current model is already overparameterized for the dataset.  Perfect fit for a smaller version of the same architecture, or trying it out on a bigger dataset.

PAI_cvpksocf

**Comment by aakanksha-singh-hub** [comment]
Hi Rorry, thanks a lot for the detailed feedback and for pushing us to look deeper. You were absolutely right about the overfitting and about SBERT already being near its efficiency ceiling on STS.

We spent the last day running a careful set of follow-up experiments to validate this rather than trying to optimize around it. Tracking `add_extra_score` and inspecting the PAI switching behavior confirmed that the baseline reaches peak performance almost immediately, leaving very little headroom for dendritic gains.

In hindsight, we agree that STS was not the most efficient dataset choice for demonstrating accuracy improvements, and that’s on us. Instead of forcing marginal gains, we’ve reframed the submission to focus on what the experiments do reveal: where dendritic optimization stops adding value, how overfitting and capacity saturation manifest in practice, and how PAI still behaves correctly by expanding architecture even when validation plateaus.

We’ve updated the README and case study to present this as a boundary-analysis, which we think is still useful for the community in clarifying when dendrites are (and aren’t) the right tool.

Thanks again for the guidance, it helped turn an initially misleading setup into a much more honest and informative result.

**Comment by RorryB** [comment]
Thank you for the honest assessment and for taking the time to dig deeper into this. I really appreciate you running those follow-up experiments and reframing the work as a boundary-analysis rather than trying to force results that weren't there.
This kind of transparent investigation into when and why our technique doesn't add value is certainly beneficial for the community. We'll keep the submission around for anyone who might benefit from reviewing it in the future.

Thanks!

**Comment by aakanksha-singh-hub** [comment]
Thank you so much, really appreciate the thoughtful feedback. Glad the analysis is useful, and thanks again for taking the time to review it.
